### ISSUE TITLE
"no-inferrable-types" produces wrong fixes for TypeScript >= 3.4

### ISSUE THREAD

**Issue by bowenni** [palantir/tslint#4742]
### Bug Report

- __TSLint version__: 5.16.0
- __TypeScript version__: 3.4.0
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
declare function foo(f: Function): void;
foo((p: number = 1) => null);
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "no-inferrable-types": true
  }
}
```

#### Actual behavior
Error:
`Type number trivially inferred from a number literal, remove type annotation`

If I apply the suggested fix, the `: number` part is removed and the compilation is broken https://www.typescriptlang.org/play/#src=declare%20function%20foo(f%3A%20Function)%3A%20void%3B%0D%0Afoo((p%20%3D%201)%20%3D%3E%20null)%3B (with `--noImplicitAny`).

This is because with https://github.com/Microsoft/TypeScript/issues/28816#issuecomment-457408663 TypeScript prefers contextual parameter type over initializer type. So `Function` takes the precedence over `= 1` and the `: number` type annotation becomes necessary.

#### Expected behavior
No errors.

**Comment by JoshuaKGoldberg** [comment]
What an amusing bug 😄 nice find!

Function overloads are relatively less common, overloads with different types especially so. `no-inferrable-types` right now is an untyped rule. Adding a fix for this bug would need to involve making it an `OptionallyTypedRule` so folks can still use it without type info. Seems like a rare enough edge case not to change typical usage requirements for it.

**Comment by bowenni** [comment]
I think it's debatable whether it's a bug of TSLint or TypeScript.
I'm very much with Ryan on https://github.com/microsoft/TypeScript/issues/30840#issuecomment-483918149 that the change in TS 3.4 only creates new implicit anys.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!
